src/taskgraph/transforms/matrix.py

```diff
@@ -0,0 +1,130 @@
+                    """
+                Don't perform any task substitution (default: false).
+
+                If specified, the ``task_context`` transforms **must** run afterwards.
+                """.lstrip()
+                ),
+            ): bool,
+            Optional(
+                "exclude",
+                description=dedent(
+                    """
+                Exclude the specified combination(s) of matrix values from the
+                final list of tasks.
+
+                If only a subset of the possible rows are present in the
+                exclusion rule, then *all* combinations including that subset
+                subset will be excluded.
+                """.lstrip()
+                ),
+            ): [{str: str}],
+            Optional(
+                "set-name",
+                description=dedent(
+                    """
+                Sets the task name to the specified format string.
+
+                Useful for cases where the default of joining matrix values by
+                a dash is not desired.
+                """.lstrip()
+                ),
+            ): str,
+            Optional(
+                "substitution-fields",
+                description=dedent(
+                    """
+                List of fields in the task definition to substitute matrix values into.
+
+                If not specified, all fields in the task definition will be
+                substituted.
+                """
+                ),
+            ): [str],
+            Extra: [str],
+        },
+        Extra: object,
+    },
+)
+"""Schema for matrix transforms."""
+
+transforms = TransformSequence()
+transforms.add_validate(MATRIX_SCHEMA)
+
+
+def _resolve_matrix(tasks, key, values, exclude):
+    for task in tasks:
+        for value in values:
+            new_task = deepcopy(task)
+            new_task["name"] = f"{new_task['name']}-{value}"
+
+            matrix = new_task.setdefault("attributes", {}).setdefault("matrix", {})
+            matrix[key] = value
+
+            for rule in exclude:
+                if all(matrix.get(k) == v for k, v in rule.items()):
+                    break
+            else:
+                yield new_task
+
+
+@transforms.add
+def split_matrix(config, tasks):
+    for task in tasks:
+        if "matrix" not in task:
+            yield task
+            continue
+
+        matrix = task.pop("matrix")
+        set_name = matrix.pop("set-name", None)
+        fields = matrix.pop("substitution-fields", task.keys())
+        defer = matrix.pop("defer-substitution", False)
+        exclude = matrix.pop("exclude", {})
+
+        new_tasks = [task]
+        for key, values in matrix.items():
+            new_tasks = _resolve_matrix(new_tasks, key, values, exclude)
+
+        for new_task in new_tasks:
+            if set_name:
+                if "name" not in fields:
+                    fields.append("name")
+                new_task["name"] = set_name
```

Comment:
I see that `name` is getting set both in `_resolve_matrix` and here. Can this be centralized in one place or the other? (I don't really see why `name` needs to be set when resolving the matrix, but maybe I'm missing something.)

Reply:
We could centralize it here, it would just mean we need to iterate over the matrix values again. I don't have a strong preference, want me to move it?

Reply:
Up to you!